### ISSUE TITLE
Retrieval Task Mixed Precision Dtype Issue

### ISSUE THREAD

**Issue by Pelps12** [tensorflow/recommenders#498]
For a mixed precision approach, I observed that the query_embeddings for the retrieval task gets cast to the global precision policy while the candidate_embedding preserves its dtype. It causes the linalg computation in the call method in retrieval to error due to different dtypes.
![Screenshot 2022-06-05 022000](https://user-images.githubusercontent.com/31068338/172040121-04e2bbea-bdc6-4ff3-92e9-8db1fa5fe56e.png)

**Comment by Hedluund** [comment]
I have the same problem!

**Comment by thushv89** [comment]
This is pretty annoying and has been grappling with this for days. Does anyone know, where in the code this is enforced (i.e. `query_embeddings` getting cast to the `global_policy` dtype?

One workaround is to keep the `global_policy` in `float32` and make `tf.keras.Model` objects of type `mixed_float16`. Here's an example code.

**Caveat**: I don't really know if such model live up to the performance gains obtained with mixed precision training. Nor that I have done a proper accuracy comparison yet. Nonetheless, it does work and from the `tensorboard` profiling I did, it does use `float16` on the GPU. 

```
mixed_precision.set_global_policy('float32')

class TwoTowerModel(tfrs.models.Model):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.query_model = QueryModel(dtype="mixed_float16")
        self.candidate_model = CandidateModel(dtype="mixed_float16")
        self.task = tfrs.tasks.Retrieval(...)

    def compute_loss(self, features, training=False):
        
        query_embeddings = tf.cast(self.query_model(features["query"]), tf.float32)
        item_embeddings = tf.cast(self.query_model(features["item"]), tf.float32)

        return self.task(query_embeddings=query_embeddings, candidate_embeddings=item_embeddings, compute_metrics=not training)
```

**Comment by patrickorlando** [comment]
Keras automatically casts input to layers depending on their dtype.
The issue is that the `Retrieval` task is a keras layer for which the dtype is set as the global policy. The dtype parameter can't be set for this layer curently. This issue also applies to the loss correction layers `SamplingBiasCorrection`, `RemoveAccidentalHits`, `HardNegativeMining`.

So essentially it's best to follow @thushv89 advice for now.

If desired @maciejkula, I'm happy to contribute a PR to expose the layer dtype for the task and loss layers?

**Comment by danielmejiaMCO** [comment]
Same problem here!

**Comment by sergeygeo** [comment]
+1

**Comment by TimSchmeier** [comment]
Any movement on this issue? @patrickorlando did you ever get around to opening a PR?

**Comment by TimSchmeier** [comment]
@maciejkula I'd be happy to open another PR to enable this if this project is still active.
